{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Generate sequence landscape figure.ipynb","provenance":[],"collapsed_sections":[],"machine_shape":"hm"},"kernelspec":{"display_name":"Python 2","language":"python","name":"python2"},"language_info":{"codemirror_mode":{"name":"ipython","version":2},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython2","version":"2.7.15"}},"cells":[{"cell_type":"markdown","metadata":{"colab_type":"text","id":"jPW0jt9iGHLd"},"source":["# Sequence landscape figure\n","\n","Requires: theano and biopython.\n","\n","Must be run in python 2 due to theano requirements. Other scripts used to produced required data:\n","\n","  - \"Load Taxonomy Labels from Uniprot\" (jupyter notebook) was used to produced taxonomy assignments (which is able to produce `PSE1_NATURAL_TAXONOMY.csv` and `PSEAB_NATURAL_TAXONOMY.csv`)\n","\n","Note: training of the variational autoencoder takes a long time.\n","\n","Requires `sequence_space_data.zip` from https://evcouplings.org/3Dseq\n"]},{"cell_type":"code","metadata":{"colab_type":"code","id":"idiqCVIiGHLe","colab":{}},"source":["%matplotlib inline\n","%load_ext autoreload\n","%autoreload 2\n","%config InlineBackend.figure_format='png'"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"colab_type":"code","id":"F8CWOl1KvQUG","colab":{}},"source":["PROJECT_ROOT = '.'"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"colab_type":"code","id":"ZFWbn_l1GHLh","colab":{}},"source":["import sys\n","sys.path.append(PROJECT_ROOT)\n","\n","import os, copy, time\n","import matplotlib.pyplot as plt\n","import seaborn as sns\n","import numpy as np\n","import pandas as pd\n","import scipy\n","\n","from Bio import SeqIO\n","from Bio import pairwise2\n","\n","#then import the deepseqs package code\n","import model\n","import helper\n","import train\n","\n","import theano\n","theano.config.gcc.cxxflags = \"-Wno-c++11-narrowing\" #FIXES ISSUE WITH COMPILING THEANO LIVE COMPILATION"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"colab_type":"text","id":"4ht-7o-TGHLj"},"source":["# Global Variables"]},{"cell_type":"code","metadata":{"colab_type":"code","id":"Uoy5KvYtGHLk","colab":{}},"source":["PROJECT_ROOT = '.'\n","\n","ROUNDS = 'ROUNDS'\n","QUERY = 'QUERY'\n","NATURAL_FULL = 'Natural Full'\n","NATURAL_SUBSAMPLE_RANDOM = 'Natural random'\n","RND10 = 'Round 10'\n","RND20 = 'Round 20'\n","RND2 = 'Round 2'\n","RND4 = 'Round 4'\n","RND8 = 'Round 8'\n","\n","WT_FASTANAME_TO_COMMONNAME = {\n","    'BLAT_PSE1/1-266': 'PSE1',\n","    'WT_PSEAB': 'AAC6',\n","    'WT_PSEAB/1-148': 'AAC6'\n","}\n","\n","ALIGNMENT_DIR_PSE1 = PROJECT_ROOT+'/data.PSE1'\n","ALIGNMENT_DIR_AAC6 = PROJECT_ROOT+'/data.AAC6'\n","FIGURE_OUTPUT_DIR = PROJECT_ROOT+'/figures'\n","\n","PSE1_ATTRIBUTES = {\n","    'name': 'PSE1',\n","    'sequence_start_offset': 6, #these are lower case - not aligned\n","    'sequence_end_offset': 4,   #these are lower case - not aligned\n","    'alignment_directory': ALIGNMENT_DIR_PSE1,\n","    'full_natural_alignment_filename': \n","        ALIGNMENT_DIR_PSE1+'/7fa1c5691376beab198788a726917d48_b0.4.a2m',\n","    'label_filenames': {\n","        QUERY: ALIGNMENT_DIR_PSE1+'/PSE1.fas',\n","        NATURAL_SUBSAMPLE_RANDOM: \n","            ALIGNMENT_DIR_PSE1+'/PSE1_natural_>85%ungapped_5K_subsample.a2m',\n","        RND20: ALIGNMENT_DIR_PSE1+'/Rnd20_init.fas', \n","        RND10: ALIGNMENT_DIR_PSE1+'/Rnd10.fas'\n","    },\n","    'label_color':{ #taken from \"tab10\" matplotlib colors\n","        NATURAL_FULL: 'black',\n","        NATURAL_SUBSAMPLE_RANDOM: 'black',\n","        RND10: '#3a76af',#'firebrick',\n","        RND20: '#ef8536', #'darkgreen', \n","        QUERY: 'yellow'\n","    },\n","    'taxonomy_filename': ALIGNMENT_DIR_PSE1+'/PSE1_NATURAL_TAXONOMY.csv',\n","    'vae_random':{\n","        'file_prefix': 'PSE1_2latent_5ksubsample',\n","        'alignment_filename': ALIGNMENT_DIR_PSE1+'/PSE1_natural_>85%ungapped_5K_subsample.a2m',\n","        'working_dir': ALIGNMENT_DIR_PSE1\n","    },\n","    'class_colors': { #min 100 seq members\n","       'Actinobacteria': \"#1f77b4\", 'Alphaproteobacteria': \"#ff7f0e\", 'Bacilli': \"#2ca02c\", \n","       'Betaproteobacteria': \"#d62728\", 'Gammaproteobacteria': \"#17becf\", \n","    }\n","}\n","\n","AAC6_ATTRIBUTES = {\n","    'name': 'AAC6',\n","    'sequence_start_offset': 0,\n","    'sequence_end_offset': 0,\n","    'alignment_directory': ALIGNMENT_DIR_AAC6,\n","    'full_natural_alignment_filename': \n","        ALIGNMENT_DIR_AAC6+'/44883374318b63406a7415d2f4d4cfc1_b0.4.a2m',\n","    'label_filenames': {\n","        QUERY: ALIGNMENT_DIR_AAC6+'/WT_PSEAB.fas',\n","        NATURAL_SUBSAMPLE_RANDOM: \n","            ALIGNMENT_DIR_AAC6+'/AAC6_natural_>85%ungapped_5K_subsample.a2m',\n","        RND2: ALIGNMENT_DIR_AAC6+'/G2_PSEAB_0_St115_50ksub.fas', \n","        RND4: ALIGNMENT_DIR_AAC6+'/G4_PSEAB_0_St115_50ksub.fas', \n","        RND8: ALIGNMENT_DIR_AAC6+'/G8_all_PSEAB_0_St115_50ksub.fas',\n","    },\n","    'label_color':{\n","        NATURAL_FULL: 'black',#'gray',\n","        NATURAL_SUBSAMPLE_RANDOM: 'black',#'gray',\n","        RND2: '#c53932',#'red', \n","        RND4: '#8d6bb8',#'limegreen', \n","        RND8: '#85584e',#'cornflowerblue',\n","        QUERY: 'yellow'\n","    },\n","    'taxonomy_filename': ALIGNMENT_DIR_AAC6+'/AAC6_NATURAL_TAXONOMY.csv',\n","    'vae_random':{\n","        'file_prefix': 'AAC6_2latent_5ksubsample',\n","        'alignment_filename': ALIGNMENT_DIR_AAC6+'/AAC6_natural_>85%ungapped_5K_subsample.a2m',\n","        'working_dir': ALIGNMENT_DIR_AAC6,\n","    },\n","    'class_colors': { #min 100 seq members\n","       'Actinobacteria': \"#1f77b4\", 'Alphaproteobacteria': \"#ff7f0e\", 'Bacilli': \"#2ca02c\", \n","       'Betaproteobacteria': \"#d62728\", 'Clostridia': \"#9467bd\", 'Gammaproteobacteria': \"#17becf\", \n","    }\n","}\n","\n","ALPHABET = 'ACDEFGHIKLMNPQRSTVWY'\n","AA_IDX_DICT = {aa:i for i,aa in enumerate(ALPHABET)}\n","IDX_AA_DICT = {idx: aa for aa, idx in AA_IDX_DICT.iteritems()}"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"colab_type":"text","id":"M9oIeXXTGHLm"},"source":["# Setup Data Munging Functions"]},{"cell_type":"code","metadata":{"colab_type":"code","id":"R06z7a-NGHLn","colab":{}},"source":["def getMutationCountFromQuery(query_in_binary, sequences_in_binary):\n","    '''\n","    Returns a list of the number of mutations each sequence in sequences has\n","    relative to the query sequence.\n","    '''\n","    toreturn = []\n","    for target_seq_in_binary in sequences_in_binary:\n","        num_matching_residues = np.sum( np.logical_and(\n","            query_in_binary, target_seq_in_binary\n","        ) )\n","        toreturn.append( (len(query_in_binary)/len(ALPHABET)) - num_matching_residues )\n","    return toreturn\n","\n","def encodeSequencesAsBinary(sequences, flatten_each_sequence = False):\n","    '''\n","    Encodes a set of polypeptide sequences in to binary.\n","    '''\n","    start = time.time()\n","    if type(sequences[0]) is SeqIO.SeqRecord: \n","        sequences = [str(seq.seq) for seq in sequences]\n","    \n","    seq_arr = np.zeros((len(sequences),len(sequences[0]),len(ALPHABET)))\n","    for i,seq in enumerate(sequences):\n","        for j,aa in enumerate(seq):\n","            if aa in AA_IDX_DICT: #non-nomal letters just get all zeros\n","                seq_arr[i,j,AA_IDX_DICT[aa]] = 1.\n","                \n","    if flatten_each_sequence:\n","        return seq_arr.reshape((len(sequences),-1))\n","    return seq_arr\n","\n","def loadSequenceFile(filename, remove_start=0, remove_end= 0):\n","    '''\n","    Loads sequences from a fasta file. Removes initial or trailing amino acids\n","    as needed.\n","    '''\n","    print('Loading \"{0}\" (removing {1} start and {2} end residues)'.format(\n","        filename, remove_start, remove_end\n","    ))\n","    toreturn = list(SeqIO.parse(filename, 'fasta'))\n","    if remove_start or remove_end:\n","        toremove = None \n","        if remove_end: toremove = -1*remove_end\n","        toreturn = [seq[remove_start:toremove] for seq in toreturn]\n","    return toreturn\n","\n","\n","def loadFastafileIntoDF(attributes, \n","                        filename, \n","                        label, \n","                        focus_seq_only=True,\n","                        override_remove_start = None,\n","                        override_remove_end = None):\n","    '''\n","    Load a single fastafile into a dataframe.\n","    If focus_seq_only == False, no amino acids will be removed from\n","    the sequences.\n","    '''\n","    remove_start=attributes['sequence_start_offset'] \n","    remove_end=attributes['sequence_end_offset']\n","    if focus_seq_only == False:\n","        remove_start = 0\n","        remove_end = 0\n","    if override_remove_start: remove_start = override_remove_start\n","    if override_remove_end: remove_start = override_remove_end\n","    \n","    query_sequence = loadSequenceFile(\n","        attributes['label_filenames'][QUERY], \n","        remove_start=remove_start, \n","        remove_end=remove_end\n","    )[0]\n","    \n","    toreturn = pd.DataFrame()\n","    seq_records = loadSequenceFile(\n","        filename, \n","        remove_start=remove_start, \n","        remove_end=remove_end\n","    )\n","\n","    toreturn['seq_record'] = seq_records\n","    toreturn['label']= label\n","    toreturn['color']= attributes['label_color'][label]\n","    toreturn['seq_name'] = [seq.name for seq in toreturn['seq_record']]\n","    toreturn['seq_str'] = [ str(seq.seq) for seq in toreturn['seq_record'] ]\n","    toreturn['num_aminoacids'] = [\n","        len(seq_str.replace('.','').replace('-','')) for seq_str in toreturn['seq_str']\n","    ]\n","    \n","    #encode sequences as binary\n","    toreturn['seq_binary'] = list(\n","        encodeSequencesAsBinary( toreturn['seq_record'] )\n","    )\n","    toreturn['seq_binary_flat'] = list(encodeSequencesAsBinary( \n","        toreturn['seq_record'], flatten_each_sequence=True\n","    ))\n","\n","    #mut count from the query sequence\n","    toreturn['mutcount_from_query'] = getMutationCountFromQuery(\n","        encodeSequencesAsBinary([query_sequence], True)[0],\n","        toreturn.seq_binary_flat\n","    )\n","    \n","    return toreturn\n","\n","\n","def createSubsample(attributes):\n","    '''\n","    Full natural alignments were generated by EVcouplings \n","    using a bitscore of 0.4. This method will generate a \n","    5000 sequence subsample of the full alignment using\n","    the following process\n","      1. remove any sequences that are not 85% ungapped \n","         relative to the query sequence.\n","      2. Save the file to attributes['label_filenames']['NATURAL_SUBSAMPLE_***']\n","    '''\n","    query_sequence = loadSequenceFile(\n","        attributes['label_filenames'][QUERY]\n","    )[0]\n","    \n","    tmpdf = loadFastafileIntoDF(\n","        attributes, \n","        attributes['full_natural_alignment_filename'], \n","        NATURAL_FULL,\n","        focus_seq_only=False\n","    )\n","    \n","    len_all_natural_sequences = len(tmpdf)\n","    num_aminoacids_queryseq = len(query_sequence)\n","    tmpdf = tmpdf[tmpdf.num_aminoacids >= 0.85 * num_aminoacids_queryseq]\n","    tmpdf = tmpdf[tmpdf.seq_str != str(query_sequence.seq)] #remove query sequence\n","    \n","    print('{0} seqs removed due to 85% length requirement ({1})'.format(\n","        len_all_natural_sequences - len(tmpdf), \n","        0.85 * num_aminoacids_queryseq\n","    ))\n","    \n","    filename = attributes['label_filenames'][NATURAL_SUBSAMPLE_RANDOM]\n","    tmpdf = tmpdf.sample(4999)\n","    \n","    SeqIO.write(\n","        [query_sequence] + list(tmpdf.seq_record), \n","        filename, \n","        'fasta'\n","    )\n","    print('done saving 5000 seq subsample')\n","\n","    \n","\n","def loadDataframe(attributes, \n","                  labels_list=None, \n","                  load_taxonomy=True,\n","                  override_remove_start = None,\n","                  override_remove_end = None):\n","    '''\n","    Load the dataframe for a set of sequences.\n","    \n","    Parameters:\n","        attributes:       a dictionary of attributes for the requested\n","                          dataset to load\n","        labels_list:      a list of labels to load. If None, then the\n","                          full list will be used that is specified in: \n","                             attributes['label_filenames'].keys()\n","        load_taxonomy:    whether or not to include taxonomy in the\n","                          returned dataframe.\n","    '''\n","    if os.path.isfile(attributes['label_filenames'][NATURAL_SUBSAMPLE_RANDOM]) == False:\n","        createSubsample(attributes)\n","    \n","    toreturn_df = pd.DataFrame()\n","    \n","    if labels_list == None: labels_list = attributes['label_filenames'].keys()\n","    for lbl in labels_list:\n","        filename = None\n","        if lbl == NATURAL_FULL:\n","            filename = attributes['full_natural_alignment_filename']\n","        else:  filename = attributes['label_filenames'][lbl]\n","        \n","        lbl_df = loadFastafileIntoDF(\n","            attributes, filename, lbl,\n","            override_remove_start=override_remove_start,\n","            override_remove_end=override_remove_end\n","        )\n","        \n","        #add to dataframe that is returned\n","        toreturn_df = toreturn_df.append( lbl_df, ignore_index=True )\n","    \n","    if not load_taxonomy: return toreturn_df\n","    \n","    #load taxonomy\n","    taxonomy = pd.read_csv(attributes['taxonomy_filename'], sep='\\t')\n","    return pd.merge(toreturn_df, taxonomy, how='left', \n","                    left_on='seq_name', right_on='seq_name')#.drop('uniprot_name', axis=1)\n"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"colab_type":"text","id":"Il7zyZ3JGHLp"},"source":["# Load Data"]},{"cell_type":"code","metadata":{"colab_type":"code","id":"UA4D41-CGHLq","colab":{}},"source":["pse1_df = loadDataframe(\n","    PSE1_ATTRIBUTES, \n","    labels_list=[QUERY, NATURAL_SUBSAMPLE_RANDOM, RND10, RND20]\n",")"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"colab_type":"code","id":"64PvBrYVGHLt","colab":{}},"source":["aac6_df = loadDataframe(\n","    AAC6_ATTRIBUTES, \n","    labels_list=[QUERY, NATURAL_SUBSAMPLE_RANDOM, RND2, RND4, RND8]\n",")"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"colab_type":"text","id":"R7FZFIebGHLw"},"source":["# lab evolution sequence space vs natural sequence space"]},{"cell_type":"code","metadata":{"colab_type":"code","id":"6aTgHzOfGHLx","colab":{}},"source":["#\n","#\n","# data munging functions\n","#\n","#\n","def getVAEDataHelper(vae_attributes):\n","    return helper.DataHelper(\n","        alignment_file=vae_attributes['alignment_filename'],\n","        working_dir=vae_attributes['working_dir'],\n","        calc_weights=False,\n","    )\n","\n","def loadVAEModel(vae_attributes):\n","    start = time.time()\n","    data_helper = getVAEDataHelper(vae_attributes)\n","    vae_model   = model.VariationalAutoencoder(\n","        data_helper,\n","        batch_size                = 100,\n","        encoder_architecture      = [1500,1500],\n","        decoder_architecture      = [100,500],\n","        n_latent                  = 2,\n","        n_patterns                = 4,\n","        warm_up                   = 0.0,\n","        convolve_patterns         = True,\n","        conv_decoder_size         = 40,\n","        logit_p                   = 0.001,\n","        sparsity                  = 'logit',\n","        encode_nonlinearity_type  = 'relu',\n","        decode_nonlinearity_type  = 'relu',\n","        final_decode_nonlinearity = 'sigmoid',\n","        output_bias               = True,\n","        final_pwm_scale           = True,\n","        working_dir               = data_helper.working_dir\n","    )\n","    print('Done loading model - took {0}s'.format(time.time()-start))\n","    return vae_model\n","\n","def loadVAEParameters(vae_attributes, vae_model):\n","    start = time.time()\n","    vae_model.load_parameters(file_prefix=vae_attributes['file_prefix'])\n","    print('Done Loading parameters - took {0}s'.format(time.time()-start))\n","    return vae_model\n","\n","def loadVAEModelFromDisk(vae_attributes):\n","    '''\n","    Shorthand to load model and parameters -- or create them if they\n","    don't exist -- and then return the loaded model.\n","    '''\n","    try:\n","        vae_model = loadVAEModel(vae_attributes)\n","        vae_model = loadVAEParameters(vae_attributes, vae_model)\n","    except IOError:\n","        print('error: unable to load vae model. please train the model first.')\n","    return vae_model\n","\n","    \n","def projectSequencesIntoVAESpace(model, sequences_df):\n","    '''\n","    Calculates the position in Z space for each sequence in\n","    the dataframe, creates a copy of the dataframe, adds \n","    those positions (z1 and z2) as a column and returns the\n","    new dataframe \n","    '''\n","    #project all sequences into Z space\n","    batch_mu, batch_log_sigma = model.recognize(\n","        list( sequences_df.seq_binary )\n","    )\n","    toreturn = sequences_df.copy()\n","    toreturn['z1'] = np.asarray(batch_mu)[:,0]\n","    toreturn['z2'] = np.asarray(batch_mu)[:,1]\n","    return toreturn\n"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"colab_type":"code","id":"KSIRcYBgGHLz","scrolled":true,"colab":{}},"source":["pse1_vae_model = loadVAEModelFromDisk(PSE1_ATTRIBUTES['vae_random'])"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"colab_type":"code","id":"4NCZi-yiGHL2","colab":{}},"source":["aac6_vae_model = loadVAEModelFromDisk(AAC6_ATTRIBUTES['vae_random'])"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"colab_type":"code","id":"UDFTJ557GHL5","colab":{}},"source":["aac6_vae_df = projectSequencesIntoVAESpace(aac6_vae_model, aac6_df)"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"colab_type":"code","id":"vtnKC9xtGHL7","colab":{}},"source":["pse1_vae_df = projectSequencesIntoVAESpace(pse1_vae_model, pse1_df)"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"colab_type":"code","id":"0J7RhgswGHL9","colab":{}},"source":["def plotSingleJointPlot(dataframe, \n","                        label, \n","                        g=None, \n","                        plot_width_height=10, \n","                        alpha=0.4,\n","                        hist_alpha = 0.2,\n","                        num_bins = 50, \n","                        dot_size = 2,\n","                        plot_histograms=True,\n","                        subsample_size=False,\n","                        color='grey'):\n","    \n","    tmpdf = dataframe\n","    if subsample_size and subsample_size < len(dataframe):\n","        tmpdf = dataframe.sample(subsample_size)\n","    \n","    \n","    if not g:\n","        g = sns.JointGrid(\n","            x=tmpdf.z1, \n","            y=tmpdf.z2,\n","            height=plot_width_height\n","        )\n","    else:\n","        g.x = tmpdf.z1\n","        g.y = tmpdf.z2\n","        \n","    g = g.plot_joint(func=plt.scatter, color=color, \n","                     s=dot_size, alpha=alpha,\n","                     label='{0} (n={1})'.format(label, len(g.x)))\n","    \n","    if plot_histograms:\n","        g.ax_marg_x.hist(\n","            tmpdf.z1,\n","            color=color,\n","            alpha = hist_alpha,\n","            bins=num_bins\n","        )\n","        g.ax_marg_y.hist(\n","            tmpdf.z2,\n","            orientation = 'horizontal',\n","            color=color,\n","            alpha = hist_alpha,\n","            bins=num_bins\n","        )\n","    return g\n","\n","def plotSequenceSpace(dataframe, \n","                      title, \n","                      orderedLabels = [RND8, RND4, RND2],\n","                      plot_width_height=12,\n","                      plot_histograms=True,\n","                      highlightQuery=True,\n","                      save_as_filenames=None,\n","                      save_as_dpi=300,\n","                      alpha=0.4,\n","                      hist_alpha=0.2,\n","                      dot_size=2,\n","                      subsample_size=False,\n","                      color_lineage = None,   #color_lineage accepts \"phylum\" or \"class\"\n","                      lineage_colors_hm = None,\n","                      include_title=True,\n","                      include_legend=True,\n","                      override_num_y_bins = None,\n","                      override_num_x_bins = None,\n","                      query_dot_size=100):\n","    g = None\n","    if color_lineage and color_lineage not in ['phylum', 'class']:\n","        print('warning: color_lineage must be either \"phylum\" or \"class\" NOT {0}'.format(color_lineage))\n","        color_lineage = None\n","    \n","    for label in orderedLabels:\n","        #print('label == {0}'.format(label))\n","        \n","        color = list(dataframe[dataframe.label==label].color)[0]\n","        if color_lineage and label in [NATURAL_FULL, \n","                                       NATURAL_SUBSAMPLE_RANDOM]:\n","            \n","            lbl_df = dataframe[dataframe.label == label]\n","            lbl_lineage_df = lbl_df[pd.isna(lbl_df[color_lineage]) == False]\n","            if subsample_size and subsample_size < len(lbl_lineage_df):\n","                #subsample the label, not the color\n","                lbl_lineage_df = lbl_lineage_df.sample( subsample_size )\n","            \n","            coloring_hm = lineage_colors_hm\n","            for classorphylum in coloring_hm:\n","                color = coloring_hm[classorphylum]\n","                tmpdf = lbl_lineage_df[lbl_lineage_df[color_lineage] == classorphylum]\n","                \n","                g = plotSingleJointPlot(\n","                    tmpdf, classorphylum, g=g, \n","                    plot_width_height=plot_width_height,\n","                    plot_histograms=plot_histograms,\n","                    alpha=alpha, hist_alpha=hist_alpha,\n","                    color = color, dot_size=dot_size\n","                )\n","        else:\n","            g = plotSingleJointPlot(\n","                dataframe[dataframe.label==label], label, g=g, \n","                plot_width_height=plot_width_height,\n","                plot_histograms=plot_histograms,\n","                alpha=alpha, hist_alpha=hist_alpha,\n","                subsample_size=subsample_size,\n","                color = color, dot_size=dot_size\n","            )\n","    if include_title:\n","        g.fig.suptitle(title, fontsize=16, y=1.01)\n","    \n","    g.ax_joint.xaxis.get_label().set_fontsize(36)\n","    g.ax_joint.yaxis.get_label().set_fontsize(36)\n","    [ticklbl.set_fontsize(28) for ticklbl in g.ax_joint.xaxis.get_ticklabels()]\n","    [ticklbl.set_fontsize(28) for ticklbl in g.ax_joint.yaxis.get_ticklabels()]\n","    \n","    if override_num_y_bins is not None:\n","        if override_num_y_bins != 0:\n","            plt.locator_params(axis='y', nbins=override_num_y_bins)\n","        else: plt.yticks([], [])\n","\n","    if override_num_x_bins is not None:\n","        if override_num_x_bins != 0:\n","            plt.locator_params(axis='x', nbins=override_num_x_bins)\n","        else: plt.xticks([], [])\n","    \n","    if highlightQuery:\n","        query = dataframe[dataframe.label == QUERY]\n","        if len(query) != 1:\n","            print('Error: QUERY label occurs {0} times (expected only 1 occurance)'.format(\n","                len(query)\n","            ))\n","        else:\n","            g.x=query.z1\n","            g.y=query.z2\n","            g.plot_joint(\n","                plt.scatter, \n","                s=query_dot_size, \n","                color='yellow', \n","                edgecolor='black', \n","                label=WT_FASTANAME_TO_COMMONNAME[list(query.seq_name)[0]]\n","            )\n","    if include_legend:\n","        g.ax_joint.legend()\n","    \n","    if save_as_filenames:\n","        if isinstance(save_as_filenames, list):\n","            for filename in save_as_filenames:\n","                g.savefig(filename, dpi=save_as_dpi)\n","        else: \n","            g.savefig(filename, dpi=save_as_dpi)\n"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"colab_type":"code","id":"hcD-ynNdGHMK","scrolled":false,"colab":{}},"source":["plotSequenceSpace(\n","    pse1_vae_df, \n","    'PSE1 All Sequences\\ncolored by class', \n","    orderedLabels=[NATURAL_SUBSAMPLE_RANDOM, RND20, RND10],\n","    color_lineage='class',\n","    lineage_colors_hm=PSE1_ATTRIBUTES['class_colors'],\n","    save_as_filenames = [\n","        #FIGURE_OUTPUT_DIR+'/PSE1_VAE_trainnatural5ksubsample_showall_classcolor.pdf'\n","    ],\n","    save_as_dpi=300,\n","    alpha=1,\n","    hist_alpha=0.4,\n","    dot_size=10,\n","    plot_histograms=False,\n","    subsample_size=10000,\n","    include_legend=False,\n","    include_title=False,\n","    override_num_y_bins=4,\n","    override_num_x_bins=4\n",")\n","plotSequenceSpace(\n","    pse1_vae_df, \n","    'PSE1 Lab Sequences', \n","    orderedLabels=[RND20, RND10],\n","    save_as_filenames = [\n","        #FIGURE_OUTPUT_DIR+'/PSE1_VAE_trainnatural5ksubsample_showlab.pdf'\n","    ],\n","    save_as_dpi=300,\n","    alpha=1,\n","    hist_alpha=0.4,\n","    dot_size=5,\n","    subsample_size=10000,\n","    include_legend=False,\n","    include_title=False,\n","    override_num_y_bins=4,\n","    override_num_x_bins=4,\n","    query_dot_size=350\n",")"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"colab_type":"code","id":"my8eAa82GHMM","scrolled":false,"colab":{}},"source":["plotSequenceSpace(\n","    aac6_vae_df, \n","    'AAC6 All Sequences\\ncolored by class', \n","    orderedLabels=[NATURAL_SUBSAMPLE_RANDOM, RND8, RND4, RND2],\n","    color_lineage='class',\n","    lineage_colors_hm=AAC6_ATTRIBUTES['class_colors'],\n","    save_as_filenames = [\n","        #FIGURE_OUTPUT_DIR+'/AAC6_VAE_trainnatural5ksubsample_showall_colorclass.pdf'\n","    ],\n","    save_as_dpi=300,\n","    alpha=1,\n","    hist_alpha=0.4,\n","    dot_size=10,\n","    subsample_size=10000,\n","    include_legend=False,\n","    include_title=False,\n","    override_num_y_bins=4,\n","    override_num_x_bins=4\n",")\n","plotSequenceSpace(\n","    aac6_vae_df, \n","    'AAC6 Lab Sequences', \n","    orderedLabels=[RND8, RND4, RND2],\n","    save_as_filenames = [\n","        #FIGURE_OUTPUT_DIR+'/AAC6_VAE_trainnatural5ksubsample_showlab.pdf'\n","    ],\n","    save_as_dpi=300,\n","    alpha=1,\n","    hist_alpha=0.4,\n","    dot_size=5,\n","    subsample_size=10000,\n","    include_legend=False,\n","    include_title=False,\n","    override_num_y_bins=4,\n","    override_num_x_bins=4,\n","    query_dot_size=350\n",")"],"execution_count":0,"outputs":[]}]}